genl_report.c

```diff
@@ -0,0 +1,155 @@
+    } else {
+        rcu_read_unlock();
+        return -ENODEV;
+    }
+
+    urr = find_urr_by_id(gtp, seid, urr_id);
+    if (!urr) {
+        rcu_read_unlock();
+        return -ENOENT;
+    }
+
+    skb_ack = genlmsg_new(NLMSG_GOODSIZE, GFP_ATOMIC);
+    if (!skb_ack) {
+        rcu_read_unlock();
+        return -ENOMEM;
+    }
+
+    urr->end_time = ktime_get_real();
+    err = gtp5g_genl_fill_usage_report(skb_ack,
+            NETLINK_CB(skb).portid,
+            info->snd_seq,
+            info->nlhdr->nlmsg_type,
+            urr);
+    urr->start_time = ktime_get_real();
+
+    if (err) {
+        kfree_skb(skb_ack);
+        rcu_read_unlock();
+        return err;
+    }
+    rcu_read_unlock();
+
+    return genlmsg_unicast(genl_info_net(info), skb_ack, info->snd_portid); 
+}
+
+
+static int gtp5g_genl_fill_volume_measurement(struct sk_buff *skb, struct urr *urr)
+{
+    struct nlattr *nest_volume_measurement;
+
+    nest_volume_measurement = nla_nest_start(skb, GTP5G_UR_VOLUME_MEASUREMENT);
+    if (!nest_volume_measurement)
+        return -EMSGSIZE;
+
+    if (nla_put_u8(skb, GTP5G_UR_VOLUME_MEASUREMENT_FLAGS, urr->volmeasurement.flag))
+        return -EMSGSIZE;
+    if (nla_put_u64_64bit(skb, GTP5G_UR_VOLUME_MEASUREMENT_TOVOL, urr->volmeasurement.totalVolume , 0))
+        return -EMSGSIZE;
+    if (nla_put_u64_64bit(skb, GTP5G_UR_VOLUME_MEASUREMENT_UVOL, urr->volmeasurement.uplinkVolume, 0))
+        return -EMSGSIZE;
+    if (nla_put_u64_64bit(skb, GTP5G_UR_VOLUME_MEASUREMENT_DVOL, urr->volmeasurement.downlinkVolume, 0))
+        return -EMSGSIZE;
+    if (nla_put_u64_64bit(skb, GTP5G_UR_VOLUME_MEASUREMENT_TOPACKET, urr->volmeasurement.totalPktNum, 0))
+        return -EMSGSIZE;
+    if (nla_put_u64_64bit(skb, GTP5G_UR_VOLUME_MEASUREMENT_UPACKET, urr->volmeasurement.uplinkPktNum, 0))
+        return -EMSGSIZE;
+    if (nla_put_u64_64bit(skb, GTP5G_UR_VOLUME_MEASUREMENT_DPACKET, urr->volmeasurement.downlinkPktNum, 0))
+        return -EMSGSIZE;
+    nla_nest_end(skb, nest_volume_measurement);
+
+    urr->volmeasurement = (struct VolumeMeasurement){};
+    
+    urr->threshold_tovol -= urr->volmeasurement.totalVolume;
+    urr->threshold_uvol -= urr->volmeasurement.uplinkVolume;
+    urr->threshold_dvol -= urr->volmeasurement.downlinkVolume;
+
+    return 0;
+}
+
+int gtp5g_genl_fill_usage_report(struct sk_buff *skb, u32 snd_portid, u32 snd_seq,
+    u32 type, struct urr *urr)
+{
+    void *genlh;
+    genlh = genlmsg_put(skb, snd_portid, snd_seq,
+            &gtp5g_genl_family, 0, type);
+    if (!genlh)
+        goto genlmsg_fail;
+
+    /* urrid, usagereporttrigger, queryurrreference*/
+    if (nla_put_u32(skb, GTP5G_UR_URRID, urr->id))
+        goto genlmsg_fail;
+    if (nla_put_u32(skb, GTP5G_UR_URSEQN, 0))
+        goto genlmsg_fail;
+    if(nla_put_u32(skb, GTP5G_UR_QUERY_URR_REFERENCE, 0))
+        goto genlmsg_fail;
+    if(gtp5g_genl_fill_volume_measurement(skb,urr))
```

Comment:
```c
if(gtp5g_genl_fill_volume_measurement(skb, urr))
```

Reply:
I do not understand why you mark these codes, can you make some comments to explain to me?

Reply:
Need a space after comma for easy identification.

It may not be an important thing but learning morphology about programming language might be worth.

Reply:
Thanks for notifying me. I'll check the code again to maintain morphology.